extension/httpfs/httpfs_client.cpp

```diff
@@ -2,146 +2,499 @@
 #include "http_state.hpp"
 
 #define CPPHTTPLIB_OPENSSL_SUPPORT
-#include "httplib.hpp"
+
+#include <curl/curl.h>
+#include <sys/stat.h>
+#include "duckdb/common/exception/http_exception.hpp"
 
 namespace duckdb {
 
+// we statically compile in libcurl, which means the cert file location of the build machine is the
+// place curl will look. But not every distro has this file in the same location, so we search a
+// number of common locations and use the first one we find.
+static std::string certFileLocations[] = {
+	// Arch, Debian-based, Gentoo
+	"/etc/ssl/certs/ca-certificates.crt",
+	// RedHat 7 based
+	"/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem",
+	// Redhat 6 based
+	"/etc/pki/tls/certs/ca-bundle.crt",
+	// OpenSUSE
+	"/etc/ssl/ca-bundle.pem",
+	// Alpine
+	"/etc/ssl/cert.pem"};
+
+
+//! Grab the first path that exists, from a list of well-known locations
+static std::string SelectCURLCertPath() {
+	for (std::string &caFile : certFileLocations) {
+		struct stat buf;
+		if (stat(caFile.c_str(), &buf) == 0) {
+			return caFile;
+		}
+	}
+	return std::string();
+}
+
+static std::string cert_path = SelectCURLCertPath();
+
+static size_t RequestWriteCallback(void *contents, size_t size, size_t nmemb, void *userp) {
+	size_t totalSize = size * nmemb;
+	std::string* str = static_cast<std::string*>(userp);
+	str->append(static_cast<char*>(contents), totalSize);
+	return totalSize;
+}
+
+static size_t RequestHeaderCallback(void *contents, size_t size, size_t nmemb, void *userp) {
+	size_t totalSize = size * nmemb;
+	std::string header(static_cast<char*>(contents), totalSize);
+	HeaderCollector* header_collection = static_cast<HeaderCollector*>(userp);
+
+	// Trim trailing \r\n
+	if (!header.empty() && header.back() == '\n') {
+		header.pop_back();
+		if (!header.empty() && header.back() == '\r') {
+			header.pop_back();
+		}
+	}
+
+	// If header starts with HTTP/... curl has followed a redirect and we have a new Header,
+	// so we clear all of the current header_collection
```

Comment:
not sure I understand this, the comment suggests this is going to clear the header_collection, but then doesn't?